File: ProjectBulkProcessor/Upgrade/Processors/OptionsParser.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.IO.Abstractions;
using System.Linq;
using System.Reflection;
using System.Xml.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using ProjectBulkProcessor.Extensions;
using ProjectBulkProcessor.Upgrade.Interfaces;
using ProjectBulkProcessor.Upgrade.Models;

namespace ProjectBulkProcessor.Upgrade.Processors
{
    public class OptionsParser : IOptionsParser
    {
        private static readonly PropertyInfo[] OptionsModelProperties;
        public static readonly ImmutableDictionary<string, string> OldNewFrameworkAliasMap;
        public static readonly ImmutableDictionary<string, string> PropertyNameAssemblyInfoAttributeMap;
        public static readonly ImmutableDictionary<string, string> AssemblyInfoAttributePropertyNameMap;

        static OptionsParser()
        {
            AssemblyInfoAttributePropertyNameMap = new Dictionary<string, string>
            {
                [nameof(AssemblyCompanyAttribute)] = nameof(OptionsModel.Company),
                [nameof(AssemblyCopyrightAttribute)] = nameof(OptionsModel.Copyright),
                [nameof(AssemblyDescriptionAttribute)] = nameof(OptionsModel.Description),
                [nameof(AssemblyProductAttribute)] = nameof(OptionsModel.Product),
                [nameof(AssemblyVersionAttribute)] = nameof(OptionsModel.Version)
            }.ToImmutableDictionary();

            PropertyNameAssemblyInfoAttributeMap = AssemblyInfoAttributePropertyNameMap.ToImmutableDictionary(kvp => kvp.Value, kvp => kvp.Key);

            OldNewFrameworkAliasMap = new Dictionary<string, string>
            {
                // TODO: add other frameworks
                ["v4.6.2"] = "net462",
                ["v4.6.1"] = "net461",
                ["v4.5.2"] = "net452",
                ["v4.5"] = "net45",
                ["v4.0"] = "net40",
            }.ToImmutableDictionary();

            OptionsModelProperties = typeof(OptionsModel).GetProperties(BindingFlags.Public | BindingFlags.Instance);
        }

        public OptionsModel ParseProjectOptions(FileInfoBase projectFile)
        {
            var options = new OptionsModel();

            SetProjectOptions(projectFile, options);
            SetAssemblyInfoOptions(projectFile.Directory, options);

            return options;
        }

        private void SetAssemblyInfoOptions(DirectoryInfoBase projectDirectory, OptionsModel options)
        {
            var assemblyInfoFile = projectDirectory.GetFiles("AssemblyInfo.cs", SearchOption.AllDirectories).SingleOrDefault();
            if (assemblyInfoFile == null)
            {
                return;
            }

            SyntaxTree tree;
            using (var fs = assemblyInfoFile.OpenText())
            {
                var content = fs.ReadToEnd();
                tree = CSharpSyntaxTree.ParseText(content);
            }

            var root = (CompilationUnitSyntax)tree.GetRoot();
            var attributeLists = root.AttributeLists;

            foreach (var attributeList in attributeLists)
            {
                foreach (var optionAttributeName in AssemblyInfoAttributePropertyNameMap.Keys)
                {
                    var shortAttributeName = optionAttributeName.Replace(nameof(Attribute), string.Empty);
                    var attribute = attributeList.Attributes.FirstOrDefault(a =>
                    {
                        var attributeName = (a.Name as IdentifierNameSyntax)?.Identifier.Text;
                        return string.Equals(optionAttributeName, attributeName, StringComparison.Ordinal) || string.Equals(attributeName, shortAttributeName, StringComparison.Ordinal);
                    });
                    if (attribute == null)
                    {
                        continue;
                    }

                    var property = OptionsModelProperties.Single(p => p.Name == AssemblyInfoAttributePropertyNameMap[optionAttributeName]);
                    var attributeArguments = attribute.ArgumentList.Arguments.Select(a => a.ToFullString().Trim('"'));

                    property.SetValue(options, string.Join(",", attributeArguments));
                }
            }
        }

        private void SetProjectOptions(FileInfoBase fileInfo, OptionsModel options)
        {
            XDocument doc;
            using (var fs = fileInfo.OpenRead())
            {
                doc = XDocument.Load(fs);
            }

            var oldFramework = doc.GetProjectElementByName("TargetFrameworkVersion").Value;
            if (!OldNewFrameworkAliasMap.TryGetValue(oldFramework, out var newFramework))
            {
                throw new NotSupportedException($"Framework {oldFramework} is not supported.");
            }

            options.TargetFramework = newFramework;
            options.IsExecutable = doc.GetProjectElementByName("OutputType").Value.Contains("Exe");
        }
    }
}

File: ProjectBulkProcessor.Main/ProjectUpgrade/Parsers/OptionsParser.fs
module OptionsParser

open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax
open System.Xml.Linq

type UpgradeOptions =
    { targetFramework : string
      outputType : string option
      copyright : string option
      company : string option
      authors : string option
      description : string option
      version : string option
      product : string option }
    static member defaultOptions =
        { targetFramework = "net472"
          outputType = None
          copyright = None
          company = None
          authors = None
          description = None
          version = None
          product = None }

let private getAttributeList (assemblyInfo : SyntaxTree option) =
    match assemblyInfo with
    | Some ai ->
        let root = ai.GetRoot() :?> CompilationUnitSyntax
        root.AttributeLists |> List.ofSeq |> List.collect (fun al -> al.Attributes |> List.ofSeq)
    | None -> List.empty

let private buildAttributeValue (attribute : AttributeSyntax) =
    attribute.ArgumentList.Arguments
    |> List.ofSeq
    |> List.map (fun a -> a.ToFullString().Trim '"')
    |> String.concat ","

let private buildAttributeOptions opts attributes =
    let optionBuilder optionState (attribute : AttributeSyntax) =
        match attribute.Name with
        | :? IdentifierNameSyntax as ins ->
            match ins.Identifier.Text with
            | "AssemblyCompany" -> { optionState with company = Some(buildAttributeValue attribute) }
            | "AssemblyCopyright" -> { optionState with copyright = Some(buildAttributeValue attribute) }
            | "AssemblyDescription" -> { optionState with description = Some(buildAttributeValue attribute) }
            | "AssemblyProduct" -> { optionState with product = Some(buildAttributeValue attribute) }
            | "AssemblyVersion" -> { optionState with version = Some(buildAttributeValue attribute) }
            | _ -> optionState
        | _ -> optionState
    List.fold optionBuilder opts attributes

let private buildCsprojOptions opts (xdoc : XDocument) =
    let newFramework =
        match XmlHelpers.getProjectElementByName "TargetFrameworkVersion" xdoc with
        | None -> "net471"
        | Some xml ->
            match xml.Value with
            | "v4.6.1" -> "net461"
            | "v4.6.2" -> "net462"
            | _ -> "net471"

    let outputType = XmlHelpers.getProjectElementByName "OutputType" xdoc |> Option.map (fun x -> x.Value)
    { opts with targetFramework = newFramework
                outputType = outputType }

let buildProjectOptions : SyntaxTree option -> XDocument -> UpgradeOptions =
    getAttributeList
    >> buildAttributeOptions UpgradeOptions.defaultOptions
    >> buildCsprojOptions
